File: applications/faas-migration/Event-Processing/test_event_processing.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"strings"
	"time"
)

// -----------------------------------------------------------------------------
// Input Configuration
// -----------------------------------------------------------------------------

var (
	endpoint = flag.String("endpoint", "undefined", "The endpoint to the ToDo API implementation")
	delay    = flag.Int("delay", 90, "The delay, in seconds, between insertion and validation")
	count    = flag.Int("count", 50, "Number of events per type to insert")
)

func init() {
	flag.Parse()
}

// -----------------------------------------------------------------------------
// Model Definitions
// -----------------------------------------------------------------------------

type ProcessedEvent struct {
	ID        int    `json:"ID"`
	Source    string `json:"source"`
	Timestamp int64  `json:"timestamp"`
	Message   string `json:"message"`
}

type BaseEvent struct {
	Type      string `json:"type"`
	Source    string `json:"source"`
	Timestamp int    `json:"timestamp"`
}

type TemperatureEvent struct {
	BaseEvent
	Value int `json:"value"`
}

type ForecastEvent struct {
	BaseEvent
	Forecast    int    `json:"forecast"`
	ForecastFor string `json:"forecast_for"`
	Place       string `json:"place"`
}

type StateChangeEvent struct {
	BaseEvent
	Message string `json:"message"`
}

// -----------------------------------------------------------------------------
// API Definitions
// -----------------------------------------------------------------------------

type api string

func (a api) ListEvents() ([]ProcessedEvent, error) {
	res, err := http.Get(fmt.Sprintf("%slist", a))
	if err != nil {
		return nil, err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	data, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	var output []ProcessedEvent
	err = json.Unmarshal(data, &output)
	if err != nil {
		return nil, err
	}

	return output, nil
}

func (a api) GetLatest() (*ProcessedEvent, error) {
	res, err := http.Get(fmt.Sprintf("%slatest", a))
	if err != nil {
		return nil, err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	data, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	var output []ProcessedEvent
	err = json.Unmarshal(data, &output)
	if err != nil {
		return nil, err
	}

	if len(output) != 1 {
		return nil, nil
	}

	return &output[0], nil
}

func (a api) InsertStateChangeEvent(i StateChangeEvent) error {
	data, err := json.Marshal(i)
	if err != nil {
		return err
	}

	res, err := http.Post(fmt.Sprintf("%singest", a), "application/json", bytes.NewReader(data))
	if err != nil {
		return err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	_, err = ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	return nil
}

func (a api) InsertForecastEvent(i ForecastEvent) error {
	data, err := json.Marshal(i)
	if err != nil {
		return err
	}

	res, err := http.Post(fmt.Sprintf("%singest", a), "application/json", bytes.NewReader(data))
	if err != nil {
		return err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	_, err = ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	return nil
}

func (a api) InsertTemperatureEvent(i TemperatureEvent) error {
	data, err := json.Marshal(i)
	if err != nil {
		return err
	}

	res, err := http.Post(fmt.Sprintf("%singest", a), "application/json", bytes.NewReader(data))
	if err != nil {
		return err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	_, err = ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	return nil
}

// -----------------------------------------------------------------------------
// Main Method
// -----------------------------------------------------------------------------

func main() {
	appUrl := *endpoint
	if appUrl == "undefined" {
		fmt.Println("The -endpoint Flag is required!")
		os.Exit(1)
	}
	// Append Trailing / if it is missing
	if !strings.HasSuffix(appUrl, "/") {
		appUrl += "/"
	}
	api := api(appUrl)
	api.CheckEndpoint()
	evts := api.InsertEvents()
	fmt.Printf("Waiting %d Seconds to let the events process...\n", *delay)
	for i := 1; i <= *delay; i++ {
		fmt.Printf("\r%03d Seconds Remaining....", *delay-i)
		time.Sleep(time.Second)
	}
	fmt.Println()
	api.ValidateEventsInserted(evts)
	api.TestLatest()
	fmt.Println("Test Execution Done!")
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

func (a api) CheckEndpoint() {
	fmt.Printf("Checking Connectivity to Endpoint by the List Function....")
	_, err := a.ListEvents()
	if err != nil {
		fmt.Printf("Fail!\nError: %s\n", err.Error())
		os.Exit(1)
	}
	fmt.Println("Ok!")
}

func (a api) InsertEvents() []BaseEvent {
	fmt.Println("Inserting Events...")
	evts := make([]BaseEvent, 0)
	fmt.Printf("Inserting %d Temperature Events...", *count)
	for i := 0; i < *count; i++ {
		temperatureEvent := TemperatureEvent{
			BaseEvent: BaseEvent{
				Type:      "temperature",
				Source:    GenerateRandomString(30),
				Timestamp: int(time.Now().Unix() / 1000),
			},
			Value: rand.Intn(200),
		}
		err := a.InsertTemperatureEvent(temperatureEvent)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
		evts = append(evts, temperatureEvent.BaseEvent)
	}
	fmt.Println("Ok!")
	fmt.Printf("Inserting %d Forecast Events...", *count)
	for i := 0; i < *count; i++ {
		forecastEvent := ForecastEvent{
			BaseEvent: BaseEvent{
				Type:      "forecast",
				Source:    GenerateRandomString(30),
				Timestamp: int(time.Now().Unix() / 1000),
			},
			Forecast:    rand.Intn(200),
			ForecastFor: GenerateRandomString(10),
			Place:       GenerateRandomString(10),
		}
		err := a.InsertForecastEvent(forecastEvent)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
		evts = append(evts, forecastEvent.BaseEvent)
	}
	fmt.Println("Ok!")
	fmt.Printf("Inserting %d State Change Events...", *count)
	for i := 0; i < *count; i++ {
		stateChangeEvent := StateChangeEvent{
			BaseEvent: BaseEvent{
				Type:      "state_change",
				Source:    GenerateRandomString(30),
				Timestamp: int(time.Now().Unix() / 1000),
			},
			Message: GenerateRandomString(30),
		}
		err := a.InsertStateChangeEvent(stateChangeEvent)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
		evts = append(evts, stateChangeEvent.BaseEvent)
	}
	fmt.Println("Ok!")

	return evts
}

func (a api) ValidateEventsInserted(events []BaseEvent) {
	fmt.Printf("Retrieving all events from the database...")
	evts, err := a.ListEvents()
	if err != nil {
		fmt.Printf("Fail!\nError: %s\n", err.Error())
		os.Exit(1)
	}
	fmt.Println("Ok!")
	fmt.Print("Validating that all events have been processed...")
	fail := false
	notFoundCount := 0
	for _, event := range events {
		found := false
		for _, processedEvent := range evts {
			if processedEvent.Source == event.Source {
				found = true

				if event.Type == "state_change" && !strings.Contains(processedEvent.Message, "status") {
					fmt.Printf("Fail!\nMessage Processing of State Change Message Failed\n")
					os.Exit(1)
				}
				if event.Type == "forecast" && !strings.Contains(processedEvent.Message, "Forecasted") {
					fmt.Printf("Fail!\nMessage Processing of Forecast Message Failed\n")
					os.Exit(1)
				}
				if event.Type == "temperature" && !strings.Contains(processedEvent.Message, "Temperature") {
					fmt.Printf("Fail!\nMessage Processing of Temperature Message Failed\n")
					os.Exit(1)
				}

				break
			}
		}
		if !found {
			fail = true
			notFoundCount++
		}
	}
	if fail {
		fmt.Println("Warning!")
		failPerc := float64(notFoundCount) / float64(len(events))
		fmt.Printf("%03d of %03d (%f%%) insertions were not found!\n", notFoundCount, len(events), failPerc)
		return
	}
	fmt.Println("Ok!")
}

func (a api) TestLatest() {
	fmt.Printf("Fetching latest element...")
	e, err := a.GetLatest()
	if err != nil {
		fmt.Printf("Fail!\nError: %s\n", err.Error())
		os.Exit(1)
	}
	fmt.Printf("Ok! Got ID %d\n", e.ID)
}

// -----------------------------------------------------------------------------
// Utility Methods
// -----------------------------------------------------------------------------

const randomStringCharset = "abcdefghijklmnopqrstuvwxyz"

var charsetRunes []rune

func init() {
	charsetRunes = bytes.Runes([]byte(randomStringCharset))
	rand.Seed(time.Now().Unix())
}

func GenerateRandomString(l int) string {
	id := ""
	for i := 0; i < l; i++ {
		idx := rand.Intn(len(charsetRunes))
		id += string(charsetRunes[idx])
	}
	return id
}

File: applications/faas-migration/Event-Processing/test_event_processing.py
import sys
import requests
import time

class SentEvent:
    def __init__(self, _type, _source):
        self.type = _type
        self.source = _source

class ProcessedEvent:
    def __init__(self, id, source, timestamp, message):
        self.id = id
        self.source = source
        self.timestamp = timestamp
        self.message = message

    @classmethod
    def fromjson(cls, json_object):
        return cls(json_object['ID'], json_object['source'], json_object['timestamp'], json_object['message'])

    def tojson(self):
        return {'ID': self.id, 'source': self.source, 'timestamp': self.timestamp, 'message': self.message}


class API:
    def __init__(self, endpoint):
        if endpoint[-1] == '/':
            self.endpoint = endpoint
        else:
            self.endpoint = endpoint + "/"

    # API functions

    def listEvents(self):
        resp = requests.get(url=self.endpoint + "list")
        ret = []
        if resp.status_code == 200:
            if resp.content is not None:
                data = resp.json()
                if type(data) == list:
                    for item in data:
                        ret.append(ProcessedEvent.fromjson(item))
                else:
                    ret.append(ProcessedEvent.fromjson(data))
        else:
            print("failed listEvents() with error code " + str(resp.status_code), file=sys.stderr)
            return None
        return ret

    def getLatest(self):
        resp = requests.get(url=self.endpoint + "latest")
        ret = []
        if resp.status_code == 200:
            if resp.content is not None:
                data = resp.json()
                if type(data) == list:
                    for item in data:
                        ret.append(ProcessedEvent.fromjson(item))
                else:
                    ret.append(ProcessedEvent.fromjson(data))
        else:
            print("failed getLatest() with error code " + str(resp.status_code), file=sys.stderr)
            return None
        if len(ret) == 1:
            return ret[0]
        else:
            print("No item in Latest.")
            return None

    # Tests

    def checkEndpoint(self):
        res = self.listEvents() is not None
        if res:
            print("API available!")
        else:
            print("API unavailable!")
        return res

    def insertEvents(self, spec, count):
        transfer_log_name = "transfer_log.csv"
        transfer_log_path = spec.logs_directory() / transfer_log_name

        envs = {
            'URL': self.endpoint
        }
        options = f"--logformat raw --console-output={transfer_log_path}"
        spec.run_k6(envs, options)
        # HACK to fix file permissions of transfer_log on Linux
        if spec.host_system() == 'Linux':
            spec.run('sb fix_permissions --local', image='serverless-benchmarker')

        # Read output
        events = []
        with open(transfer_log_path, "r") as transfer_log:
            for line in transfer_log:
                if line.startswith("transfer"):
                    splits = [x.strip() for x in line.split(",")]
                    events.append(SentEvent(splits[1], splits[2]))

        return events

    def validateEventsInserted(self, events):
        print("Retrieving all events from the database...")
        evts = self.listEvents()
        if evts is None:
            return None
        print("Ok!\nValidating that all events have been processed...")
        fail = False
        notFoundCount = 0
        for sent_event in events:
            found = False
            for processed_event in evts:
                if processed_event.source == sent_event.source:
                    found = True
                    if sent_event.type == "state_change" and not "status" in processed_event.message:
                        print("Fail!\nMessage Processing of State Change Message Failed", file=sys.stderr)
                        return None
                    if sent_event.type == "forecast" and not "Forecasted" in processed_event.message:
                        print("Fail!\nMessage Processing of Forecast Message Failed", file=sys.stderr)
                        return None
                    if sent_event.type == "temperature" and not "Temperature" in processed_event.message:
                        print("Fail!\nMessage Processing of Temperature Message Failed", file=sys.stderr)
                        return None
                    break
            if not found:
                fail = True
                notFoundCount = notFoundCount + 1
        if fail:
            print("Warning!\n%d of %d insertions were not found!" % (notFoundCount, len(events)))
            return None
        print("Ok!")
        return "ok!"

    def testLatest(self):
        print("Fetching latest element...")
        l = self.getLatest()
        if l is None:
            return None
        print("Ok! Got ID %d" % l.id)


def test(spec, endpoint, delay, count):
    api = API(endpoint)
    if not api.checkEndpoint():
        return None
    evts = api.insertEvents(spec, count)
    print("Waiting %d Seconds to let the events process..." % delay)
    time.sleep(delay)
    if api.validateEventsInserted(evts) is None:
        return None
    if api.testLatest() is None:
        return None
    print("Test Execution Done Successfully!")
    return "ok!"

